src/main/java/com/direwolf20/buildinggadgets/common/util/blocks/RegionSnapshot.java

```diff
@@ -0,0 +1,411 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntList;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
+import net.minecraft.block.Block;
+import net.minecraft.block.material.Material;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.NBTTagList;
+import net.minecraft.nbt.NBTUtil;
+import net.minecraft.state.IProperty;
+import net.minecraft.tileentity.TileEntity;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+import net.minecraft.world.dimension.DimensionType;
+import net.minecraftforge.common.util.BlockSnapshot;
+import net.minecraftforge.common.util.Constants;
+import net.minecraftforge.common.util.TriPredicate;
+import net.minecraftforge.fml.server.ServerLifecycleHooks;
+
+import javax.annotation.Nonnull;
+import java.util.*;
+import java.util.function.BiPredicate;
+
+public class RegionSnapshot {
+
+    private static final String DIMENSION = "dim";
+    private static final String BLOCK_FRAMES = "frames";
+    private static final String BLOCK_PALETTES = "palettes";
+    private static final String BLOCK_SNAPSHOTS = "block_snapshots";
+
+    private static NBTTagCompound serialize(NBTTagCompound tag, RegionSnapshot snapshot) {
+        tag.setInt(DIMENSION, snapshot.world.getDimension().getType().getId());
+        snapshot.region.serializeTo(tag);
+
+        // Palette serialization begin
+        NBTTagList palettes = new NBTTagList();
+        Object2IntMap<IBlockState> mapPalettes = new Object2IntOpenHashMap<>();
+        Set<IBlockState> recordedPalettes = new HashSet<>();
+
+        // "nothing should be here" block state
+        palettes.add(new NBTTagCompound());
+        // Palette serialization end
+
+        // Streaking: a way to record a number of repeating block states, using only 2 integers
+        // Storage frame: one or two numbers indicating a sequence of block states
+        // Regular frame: a type of storage frame; 1 number, representing the ID of the block state recorded
+        // Streak frame: a type of storage frame; 2 numbers, where the first one is positive, representing the ID of the repeating block state; the second one is negative, representing the number of repeating block states
+
+        // Complete storage frames, will be serialized into a NBTTagList later
+        IntList frames = new IntArrayList();
+
+        // The repeating block state, if any
+        IBlockState lastStreak = null;
+        // Number of repeating block states
+        int streak = 0;
+
+        boolean firstIteration = true;
+        for (IBlockState state : snapshot.blockStates) {
+            // Palette serialization begin
+            if (!recordedPalettes.contains(state)) {
```

Comment:
This needs a check whether we exceed the maximum amount of allowed palettes - unlikely, but theoretically possible.

Writing an error to the log and stopping further serilisation would propably be the best (we don't want to crash here, but writing an alternative format for reeeaaaallllyyyy large stuff would definitly not be worth it)